docker/loom-output/create_loom_optimus.py

```diff
@@ -0,0 +1,476 @@
+
+    Args:
+        verbose (bool): whether to output verbose messages for debugging purposes
+    """
+
+    # read the gene metrics names and values
+    input_path = args.gene_metrics
+    if input_path.endswith(".gz"):
+        with gzip.open(input_path, "rt") as f:
+            gene_metrics = [row for row in csv.reader(f)]
+    else:
+        with open(input_path, "r") as f:
+            gene_metrics = [row for row in csv.reader(f)]
+
+    if args.verbose:
+        logging.info("# gene numeric metadata {0}".format(len(gene_metrics[0][1:])))
+
+    # read the gene ids  and adds into the gene_ids dataset
+    gene_ids = np.load(args.gene_ids)
+    
+    # add the gene names for the gene ids
+    if args.annotation_file and len(gene_ids):
+        gene_id_name_map = create_gene_id_name_map(args.annotation_file)
+        gene_names = [
+            gene_id_name_map.get(gene_id, "") for gene_id in gene_ids
+        ]
+
+
+    # Gene metric values, the row and column sizes
+    gene_ids_location = {
+        gene_id: index for index, gene_id in enumerate(gene_ids)
+    }
+
+    # ignore the first line with the metric names in text
+    ncols = 0
+    gene_id_to_metric_values = {}
+    for row in gene_metrics:
+        # only consider genes that are in the count matrix
+        if not row[0] in gene_ids_location:
+            continue
+
+        row_values = []
+        for value_string in row[1:]:
+            # some of the standard deviation values do not exist for one reads matches
+            try:
+                value = np.float32(value_string)
+            except ValueError:
+                value = np.nan
+            row_values.append(value)
+        gene_id_to_metric_values[row[0]] = row_values
+
+        # note that all of these lengths are assumed to be equal and this check is already done in the pipeline
+        ncols = len(row_values)
+
+    # now insert the metrics of the cells that are in count matrix, i.e.,  the global variable "cell_ids"
+
+    gene_metric_values = []
+    for gene_id in gene_ids:
+        if gene_id in gene_id_to_metric_values:
+            gene_metric_values.append(gene_id_to_metric_values[gene_id])
+        else:
+            # if no metrics for a cell present in the count matrix then fill them with np.nans
+            gene_metric_values.append([np.nan] * ncols)
+
+    nrows = len(gene_ids)
+    
+    #Generate row attribute dictionary
+    row_attrs = { "gene_names":gene_names, "ensembl_ids": gene_ids}
+
+    X=np.array(gene_metric_values)
+    numeric_field_names = gene_metrics[0][1:]
+    for i in range(0, len(numeric_field_names)):
+        name = numeric_field_names[i]
+        data = X[:, i]
+        row_attrs[name] = data
+    if args.verbose:
+        logging.info("# of genes: {0}".format(nrows))
+        logging.info("# of gene metadate metrics: {0}".format(ncols))
+    return row_attrs
+
+def generate_col_attr(args):
+    """Converts cell metrics from the Optimus pipeline to zarr file
```

Comment:
Do you need to remove this line about the zarr file?

Reply:
Changed it :)

Reply:
Changed "zarr" to "loom"

Reply:
Still see "zarr" here

Reply:
Pushed new changes.